openscap_report/scap_results_parser/parsers/rule_parser.py

```diff
@@ -80,17 +84,98 @@ def process_rule(self, rule):
 
-        return Rule(**rule_dict)
+    @staticmethod
+    def _get_remediation_exit_code(message):
+        remediation_exit_code_prefix = "Fix execution completed and returned:"
+        if message.startswith(remediation_exit_code_prefix):
+            exit_code = message.replace(remediation_exit_code_prefix, "")
+            return int(exit_code)
+        return None
+
+    @staticmethod
+    def _get_check_engine_result(message):
+        check_engine_result_prefix = "Checking engine returns:"
+        index = message.find(check_engine_result_prefix)
+        if index:
+            return message[index:]
+        return None
+
+    @staticmethod
+    def _evaluate_and_set_result(rule_id, rules, remediation_exit_code, check_engine_result):
+        if check_engine_result is not None and "fail" in check_engine_result:
+            rules[rule_id].result = "fix unsuccessful"
+
+        if remediation_exit_code is not None and remediation_exit_code > 0:
+            rules[rule_id].result = "fix failed"
+
+    def _improve_result_of_remedied_rule(self, rule_id, rules):
+        if not rules[rule_id].messages:
+            return
+        remediation_exit_code = None
+        check_engine_result = None
+
+        for message in rules[rule_id].messages:
+            if remediation_exit_code is None:
+                remediation_exit_code = self._get_remediation_exit_code(message)
+
+            if check_engine_result is None:
+                check_engine_result = self._get_check_engine_result(message)
+
+        self._evaluate_and_set_result(rule_id, rules, remediation_exit_code, check_engine_result)
+
+    @staticmethod
+    def _add_message_about_oval(rule_id, rules):
+        if "fix" in rules[rule_id].result:
+            msg = "The OVAL graph of the rule as it was displayed before the fix was performed."
+            rules[rule_id].messages.append(msg)
+
+    def _insert_rules_results(self, rules):
+        rules_results = self.test_results.findall('.//xccdf:rule-result', NAMESPACES)
+        for rule_result in rules_results:
+            rule_id = rule_result.get('idref')
+            rules[rule_id].time = rule_result.get('time')
+            rules[rule_id].result = rule_result.find('.//xccdf:result', NAMESPACES).text
+
+            messages = rule_result.findall('.//xccdf:message', NAMESPACES)
+            if messages is not None:
+                rules[rule_id].messages = []
+                for message in messages:
+                    rules[rule_id].messages.append(message.text)
+                if not rules[rule_id].messages:
+                    rules[rule_id].messages = None
+                self._improve_result_of_remedied_rule(rule_id, rules)
+            self._add_message_about_oval(rule_id, rules)
+
+    def get_rules(self):
+        rules = {}
+        for rule in self.root.findall(".//xccdf:Rule", NAMESPACES):
+            rule = self.process_rule(rule)
```

Comment:
This kind of shadowing could lead to a lot of troubles in a dynamically typed language. If you can have a Rule class entity and a rule xml sub-document it's better to introduce a naming scheme which won't allow you to mix them up. For example for all xml entities it could be `_el`s (`rule_el`).